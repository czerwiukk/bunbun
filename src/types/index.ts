export interface IBoard {
  id: number;
  name: string;
  created_at: string;
}

export interface IBoardTask {
  id: number;
  name: string;
  description: string;
  created_at: string;
  column_id: number;
}

export interface IBoardColumn {
  id: number;
  name: string;
  created_at: string;
}
