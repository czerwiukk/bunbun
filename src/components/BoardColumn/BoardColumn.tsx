import Fa from "solid-fa";
import { Component, For } from "solid-js";
import { TaskTile } from "../TaskTile/TaskTile";
import { faPlus, faTrash } from "@fortawesome/free-solid-svg-icons";
import { IBoardTask, type IBoardColumn } from "~/types";
import { createDroppable } from "@thisbeyond/solid-dnd";
import { changeColumnName } from "~/api";
import { usePendingChanges } from "~/hooks";
import { Modal } from "..";

interface BoardColumnProps {
  column: IBoardColumn;
  tasks: IBoardTask[];
  onDeleteColumn: (id: number) => void;
}

export const BoardColumn: Component<BoardColumnProps> = (props) => {
  const droppable = createDroppable(props.column.id);

  let input: HTMLInputElement;

  const handleColumnNameEdit = async (e: KeyboardEvent) => {
    if (e.key !== "Enter" || !input.value) return;
    await changeColumnName(props.column.id, input.value);
  };

  const { isColumnReloading } = usePendingChanges;

  return (
    <div
      // @ts-ignore
      use:droppable
      class="droppable mb-4"
      classList={{ "!droppable-accept": droppable.isActiveDroppable }}
    >
      <input
        id={`${props.column.id}-column-name`}
        // @ts-ignore
        ref={input}
        class="cursor-text rounded-md bg-transparent text-lg text-stone-700 outline-none dark:text-stone-100"
        value={props.column.name}
        onKeyDown={handleColumnNameEdit}
      />

      <section
        class="flex h-full w-72 flex-col justify-between rounded-md border border-stone-300 bg-stone-200 p-2 dark:border-stone-900 dark:bg-stone-800"
        classList={{ "animate-pulse": isColumnReloading(props.column.id) }}
      >
        <ul>
          <For each={props.tasks ?? []}>
            {(task) => <TaskTile task={task} />}
          </For>
        </ul>

        <div class="flex justify-between">
          <button
            class="cursor-pointer p-2 text-stone-400 transition-colors hover:text-red-600"
            onClick={() => props.onDeleteColumn(props.column.id)}
          >
            <Fa icon={faTrash} />
          </button>

          <Modal
            trigger={<Fa icon={faPlus} />}
            triggerClass="w-8 h-8 rounded-full p-2 bg-stone-300 dark:bg-stone-600 hover:bg-stone-400 hover:bg-stone-500 dark:text-stone-200 cursor-pointer grid place-content-center transition-colors"
          >
            Add task modal
          </Modal>
        </div>
      </section>
    </div>
  );
};
