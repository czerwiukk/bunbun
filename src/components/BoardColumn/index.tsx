import Fa from "solid-fa";
import { Component, For } from "solid-js";
import { TaskTile } from "../TaskTile";
import { faPlus, faTrash } from "@fortawesome/free-solid-svg-icons";
import { IBoardTask, type IBoardColumn } from "~/types";
import { createDroppable } from "@thisbeyond/solid-dnd";
import { changeColumnName } from "~/api";
import { usePendingChanges } from "~/hooks";

interface BoardColumnProps {
  column: IBoardColumn;
  tasks: IBoardTask[];
  onDeleteColumn: (id: number) => void;
}

export const BoardColumn: Component<BoardColumnProps> = (props) => {
  const droppable = createDroppable(props.column.id);

  let input: HTMLInputElement;

  const handleColumnNameEdit = async (e: KeyboardEvent) => {
    if (e.key !== "Enter" || !input.value) return;
    await changeColumnName(props.column.id, input.value);
  };

  const { isColumnReloading } = usePendingChanges;

  return (
    <div
      // @ts-ignore
      use:droppable
      class="droppable mb-4"
      classList={{ "!droppable-accept": droppable.isActiveDroppable }}
    >
      <input
        // @ts-ignore
        ref={input}
        class="text-lg text-stone-700 cursor-text outline-none focus:bg-stone-50 rounded-md"
        value={props.column.name}
        onKeyDown={handleColumnNameEdit}
      />

      <section
        class="w-72 h-full bg-stone-200 rounded-md p-4 flex flex-col justify-between border border-stone-300"
        classList={{ "animate-pulse": isColumnReloading(props.column.id) }}
      >
        <ul>
          <For each={props.tasks ?? []}>
            {(task) => <TaskTile task={task} />}
          </For>
        </ul>

        <div class="flex justify-between">
          <button
            class="text-stone-400 p-2 hover:text-red-600 transition-colors cursor-pointer"
            onClick={() => props.onDeleteColumn(props.column.id)}
          >
            <Fa icon={faTrash} />
          </button>

          <button class="w-8 h-8 rounded-full p-2 bg-stone-300 cursor-pointer grid place-content-center transition-colors hover:bg-stone-400">
            <Fa icon={faPlus} />
          </button>
        </div>
      </section>
    </div>
  );
};
