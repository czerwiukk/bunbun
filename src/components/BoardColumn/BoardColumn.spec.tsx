import { render } from "solid-testing-library";
import { describe, expect, it, vi } from "vitest";
import { IBoardColumn } from "~/types";
import { BoardColumn } from "./BoardColumn";

const column: IBoardColumn = {
  id: 1,
  name: "Test",
  created_at: "2021-08-01T12:00:00.000Z",
};

vi.mock("@thisbeyond/solid-dnd", () => ({
  createDraggable: () => () => ({}),
  createDroppable: () => () => ({}),
}));

describe("BoardColumn", () => {
  it("matches snapshot", () => {
    expect(
      render(() => (
        <BoardColumn column={column} tasks={[]} onDeleteColumn={() => null} />
      ))
    ).toMatchSnapshot();
  });
});
