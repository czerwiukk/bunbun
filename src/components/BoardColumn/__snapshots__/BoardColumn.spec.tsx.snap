// Vitest Snapshot v1

exports[`BoardColumn > matches snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="droppable mb-4"
      >
        <input
          class="text-lg text-stone-700 dark:text-stone-100 cursor-text outline-none bg-transparent rounded-md"
          id="1-column-name"
        />
        <section
          class="w-72 h-full bg-stone-200 dark:bg-stone-800 rounded-md p-2 flex flex-col justify-between border border-stone-300 dark:border-stone-900"
        >
          <ul />
          <div
            class="flex justify-between"
          >
            <button
              class="text-stone-400 p-2 hover:text-red-600 transition-colors cursor-pointer"
            >
              <svg
                aria-hidden="true"
                class="fa"
                style="height: 1em; vertical-align: -.125em; transform-origin: center; overflow: visible;"
                viewBox="0 0 448 512"
                xmlns="http://www.w3.org/2000/svg"
              >
                <g
                  transform="translate(224 256)"
                  transform-origin="112 0"
                >
                  <g
                    transform="translate(0,0) scale(1,1)"
                  >
                    <path
                      d="M135.2 17.7L128 32H32C14.3 32 0 46.3 0 64S14.3 96 32 96H416c17.7 0 32-14.3 32-32s-14.3-32-32-32H320l-7.2-14.3C307.4 6.8 296.3 0 284.2 0H163.8c-12.1 0-23.2 6.8-28.6 17.7zM416 128H32L53.2 467c1.6 25.3 22.6 45 47.9 45H346.9c25.3 0 46.3-19.7 47.9-45L416 128z"
                      fill="currentColor"
                      transform="translate(-224 -256)"
                    />
                  </g>
                </g>
              </svg>
            </button>
            <label
              class="w-8 h-8 rounded-full p-2 bg-stone-300 dark:bg-stone-600 hover:bg-stone-400 hover:bg-stone-500 dark:text-stone-200 cursor-pointer grid place-content-center transition-colors"
              for="cl-0"
            >
              <svg
                aria-hidden="true"
                class="fa"
                style="height: 1em; vertical-align: -.125em; transform-origin: center; overflow: visible;"
                viewBox="0 0 448 512"
                xmlns="http://www.w3.org/2000/svg"
              >
                <g
                  transform="translate(224 256)"
                  transform-origin="112 0"
                >
                  <g
                    transform="translate(0,0) scale(1,1)"
                  >
                    <path
                      d="M256 80c0-17.7-14.3-32-32-32s-32 14.3-32 32V224H48c-17.7 0-32 14.3-32 32s14.3 32 32 32H192V432c0 17.7 14.3 32 32 32s32-14.3 32-32V288H400c17.7 0 32-14.3 32-32s-14.3-32-32-32H256V80z"
                      fill="currentColor"
                      transform="translate(-224 -256)"
                    />
                  </g>
                </g>
              </svg>
            </label>
            <input
              class="modal-toggle"
              id="cl-0"
              type="checkbox"
            />
            <label
              class="modal cursor-pointer"
              for="cl-0"
            >
              <label
                class="modal-box relative"
              >
                Add task modal
              </label>
            </label>
          </div>
        </section>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="droppable mb-4"
    >
      <input
        class="text-lg text-stone-700 dark:text-stone-100 cursor-text outline-none bg-transparent rounded-md"
        id="1-column-name"
      />
      <section
        class="w-72 h-full bg-stone-200 dark:bg-stone-800 rounded-md p-2 flex flex-col justify-between border border-stone-300 dark:border-stone-900"
      >
        <ul />
        <div
          class="flex justify-between"
        >
          <button
            class="text-stone-400 p-2 hover:text-red-600 transition-colors cursor-pointer"
          >
            <svg
              aria-hidden="true"
              class="fa"
              style="height: 1em; vertical-align: -.125em; transform-origin: center; overflow: visible;"
              viewBox="0 0 448 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <g
                transform="translate(224 256)"
                transform-origin="112 0"
              >
                <g
                  transform="translate(0,0) scale(1,1)"
                >
                  <path
                    d="M135.2 17.7L128 32H32C14.3 32 0 46.3 0 64S14.3 96 32 96H416c17.7 0 32-14.3 32-32s-14.3-32-32-32H320l-7.2-14.3C307.4 6.8 296.3 0 284.2 0H163.8c-12.1 0-23.2 6.8-28.6 17.7zM416 128H32L53.2 467c1.6 25.3 22.6 45 47.9 45H346.9c25.3 0 46.3-19.7 47.9-45L416 128z"
                    fill="currentColor"
                    transform="translate(-224 -256)"
                  />
                </g>
              </g>
            </svg>
          </button>
          <label
            class="w-8 h-8 rounded-full p-2 bg-stone-300 dark:bg-stone-600 hover:bg-stone-400 hover:bg-stone-500 dark:text-stone-200 cursor-pointer grid place-content-center transition-colors"
            for="cl-0"
          >
            <svg
              aria-hidden="true"
              class="fa"
              style="height: 1em; vertical-align: -.125em; transform-origin: center; overflow: visible;"
              viewBox="0 0 448 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <g
                transform="translate(224 256)"
                transform-origin="112 0"
              >
                <g
                  transform="translate(0,0) scale(1,1)"
                >
                  <path
                    d="M256 80c0-17.7-14.3-32-32-32s-32 14.3-32 32V224H48c-17.7 0-32 14.3-32 32s14.3 32 32 32H192V432c0 17.7 14.3 32 32 32s32-14.3 32-32V288H400c17.7 0 32-14.3 32-32s-14.3-32-32-32H256V80z"
                    fill="currentColor"
                    transform="translate(-224 -256)"
                  />
                </g>
              </g>
            </svg>
          </label>
          <input
            class="modal-toggle"
            id="cl-0"
            type="checkbox"
          />
          <label
            class="modal cursor-pointer"
            for="cl-0"
          >
            <label
              class="modal-box relative"
            >
              Add task modal
            </label>
          </label>
        </div>
      </section>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "unmount": [Function],
}
`;
